site/en/r2/tutorials/generative/cvae.ipynb

```diff
@@ -370,11 +376,13 @@
         "  logqz_x = log_normal_pdf(z, mean, logvar)\n",
         "  return -tf.reduce_mean(logpx_z + logpz - logqz_x)\n",
         "\n",
+        "@tf.function \n"
         "def compute_gradients(model, x):\n",
         "  with tf.GradientTape() as tape:\n",
         "    loss = compute_loss(model, x)\n",
         "  return tape.gradient(loss, model.trainable_variables), loss\n",
         "\n",
+        "@tf.function \n"
```

Comment:
tf.function on compute_gradients is all that is necessary to provide the speedup.

Also, can you merge apply_gradients with compute_gradients so that there is only one tf.function required?

Reply:
So compute_loss() does not have any performance gain from @tf.function? because it is called separately in the training process.

I tested how the recursiveness of autograph works at this colab https://colab.research.google.com/drive/16zrshCFCVB-b4zIF_Z1GtN8YahuSbAib 

and it can be seen that the speed up only occurs if the called function is decorated. So by leaving the decorator off of the loss calculation, it will reduce performance.

I further tested the actual tutorial code and found that leaving @tf.function off from the compute_loss() slows the time by about 0.1s/epoch. Perhaps not the most important, but still useful, especially if the network/loss function were expanded or slower hardware being used.

Finally, I tested it where only the loss, compute gradient, and apply gradient functions were decorated. This increased per epoch training time by 0.05s /epoch.

All together the proposed singular decorator led to training time of 3.88s/epoch in colab, which is about 7% decreased performance. 

As to combining the functions, this actually improved performance dramatically, cutting off about 0.4s/epoch so I will include this change

Reply:
I just ran it with only one tf.function decorator on top of compute_and_apply_gradients, and it took 3.2 seconds per epoch. For the first epoch it takes longer because autograph traces the code to create the graph and also other initialization stuff happens. But after that its pretty fast.

Is this different from what you observed?

Reply:
Yes, I usually disregard the first epoch or two to ensure consistency. If you only use the one tf.function I get about 3.22s/epoch for the first 10 epochs (it increases after that). If I include the other decorators I average about 3.1s/epoch in the first 10 epochs. I think this is just due to the compute_loss function call that is being run each epoch. Not necessarily important, but it isn't negligible either, especially if the loss function were more intensive.

Reply:
SGTM